kimchi/src/prover.rs

```diff
@@ -1044,6 +1044,35 @@ where
         //~ 1. Squeeze the Fq-sponge and absorb the result with the Fr-Sponge.
         fr_sponge.absorb(&fq_sponge.digest());
 
+        //~ 1. Compute evaluations for the previous recursion challenges.
+        let polys = prev_challenges
+            .iter()
+            .map(|RecursionChallenge { chals, comm }| {
+                (
+                    DensePolynomial::from_coefficients_vec(b_poly_coefficients(chals)),
+                    comm.unshifted.len(),
+                )
+            })
+            .collect::<Vec<_>>();
+
+        //~ 1. Absorb evaluations for the previous recusion challenges.
+        let prev_challenge_digest = {
+            // Note: we absorb in a new sponge here to limit the scope in which we need the
+            // more-expensive 'optional sponge'.
+            let mut fr_sponge = EFrSponge::new(index.cs.fr_sponge_params.clone());
+            for prev_challenge in prev_challenges.iter() {
+                let evals = prev_challenge.evals(
+                    index.max_poly_size,
+                    &[zeta, zeta_omega],
+                    &[zeta_to_srs_len, zeta_omega_to_srs_len],
+                );
+                fr_sponge.absorb_multiple(&evals[0]);
+                fr_sponge.absorb_multiple(&evals[1]);
```

Comment:
so two things:

* why not absorb the challenges instead? In the interactive proof, you're sending the challenges to the verifier so this is what you need to absorb IIUC for fiat-shamir
* this is still missing the absorption of the prev_challenges.commitment

Reply:
for commitment, it's already done here: https://github.com/o1-labs/proof-systems/pull/670